File: src/main/java/ux/ToggleIconControl.java
package ux;

import actions.KeyCallback;
import actions.KeyObservable;
import processing.core.PImage;

public class ToggleIconControl extends Control {
    PImage toggledIcon; // right now only used with play / pause
    boolean iconToggled = false;
    PImage currentIcon;
    protected ToggleIconControl(Builder builder) {
        super(builder);
        this.currentIcon = icon;
        this.toggledIcon = loadIcon(builder.toggledIconName);
    }

    @Override
    public void onMouseReleased() {
        super.onMouseReleased();
        toggleIcon();
    }

    @Override
    protected PImage getIcon() {
        return iconToggled ? toggledIcon : icon;
    }

    @Override
    public void notifyKeyPress(KeyObservable o) {
        toggleIcon();
    }

    private void toggleIcon() {
        currentIcon = (iconToggled) ? icon : toggledIcon;
        iconToggled = !iconToggled;
    }

    public static class Builder extends Control.Builder {
        private final String  toggledIconName;

        public Builder(PGraphicsSupplier graphicsSupplier, KeyCallback callback, String iconName, String tooggledIcontName, int size) {
            super(graphicsSupplier, callback, iconName, size);
            this.toggledIconName = tooggledIcontName;
        }

        @Override
        public ToggleIconControl build() {
            return new ToggleIconControl(this);
        }

        @Override
        public Builder self() {
            return this;
        }

    }
}

File: src/main/kotlin/patterning/ux/panel/ToggleIconControl.kt
package patterning.ux.panel

import patterning.actions.KeyCallback
import patterning.actions.KeyObservable
import patterning.ux.Theme
import patterning.ux.informer.DrawingInfoSupplier
import processing.core.PImage
import java.util.*

class ToggleIconControl(builder: Builder) : Control(builder) {
    private var toggledIcon // right now only used with play / pause
            : PImage
    private var iconToggled = false
    private var singleMode = false
    private lateinit var currentIcon: PImage
    private var modeChangeCallback: KeyCallback

    init {
        toggledIcon = loadIcon(builder.toggledIconName)
        modeChangeCallback = builder.modeChangeCallback
        modeChangeCallback.addObserver(this)
    }

    override fun afterInit() {
        currentIcon = icon
    }

    override fun getCurrentIcon(): PImage {
        return if (iconToggled) toggledIcon else super.getCurrentIcon()
    }

    override fun notifyKeyPress(observer: KeyObservable) {
        highlightFromKeyPress()
        if (observer.invokeModeChange()) {
            toggleMode()
            if (!iconToggled) toggleIcon()
        } else toggleIcon()
    }

    override fun onMouseReleased() {
        super.onMouseReleased()
        toggleIcon()
    }

    private fun toggleIcon() {
        currentIcon = if (iconToggled) icon else toggledIcon
        iconToggled = !iconToggled
        if (singleMode && !iconToggled) {
            Timer().schedule(
                object : TimerTask() {
                    override fun run() {
                        toggleIcon()
                    }
                },
                Theme.singleModeToggleDuration
                    .toLong()
            )
        }
    }

    private fun toggleMode() {
        singleMode = !singleMode
    }

    class Builder(
        drawingInformer: DrawingInfoSupplier?,
        callback: KeyCallback?,
        val modeChangeCallback: KeyCallback,
        iconName: String?,
        val toggledIconName: String,
        size: Int
    ) : Control.Builder(drawingInformer, callback!!, iconName!!, size) {
        override fun self(): Builder {
            return this
        }

        override fun build(): ToggleIconControl {
            return ToggleIconControl(this)
        }
    }
}